src/Sylius/Bundle/CoreBundle/Validator/Constraints/TranslationForExistingLocalesValidator.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Validator\Constraints;
+
+use Sylius\Component\Locale\Model\LocaleInterface;
+use Sylius\Component\Product\Model\ProductVariantTranslationInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+final class TranslationForExistingLocalesValidator extends ConstraintValidator
+{
+    public function __construct(private RepositoryInterface $localeRepository)
+    {
+    }
+
+    public function validate($value, Constraint $constraint): void
+    {
+        Assert::isInstanceOf($constraint, TranslationForExistingLocales::class);
+
+        /** @var LocaleInterface[] $locales */
+        $locales = $this->localeRepository->findAll();
+
+        if (empty($locales)) {
+            return;
+        }
+
+        $localeCodes = array_map(fn (LocaleInterface $locale) => $locale->getCode(), $locales);
+
+        $translations = $value->getTranslations();
+
+        /** @var ProductVariantTranslationInterface $translation */
+        foreach ($translations as $key => $translation) {
+            if (!in_array($translation->getLocale(), $localeCodes, true)) {
+                $this->context->buildViolation($constraint->message)
+                    ->setParameter('%locales%', implode(', ', $localeCodes))
+                    ->atPath(sprintf('translations[%s]', $key))
+                    ->addViolation()
+                ;
+            }
+        }
```

Comment:
Are you testing the use of this validator somewhere?

Reply:
I tried to test it with behat and when I found out that it was not possible, I skipped writing a test unit for it, my mistake.

I added units for both creation and updating:
https://github.com/Sylius/Sylius/pull/14837/files#diff-524baa0977fcaad8462df7499d18d38d51ade64bdee67cc80ac45c20423ef51aR262-R327